dcrtimed/dcrtimed.go

```diff
@@ -441,6 +466,55 @@ func (d *DcrtimeStore) verify(w http.ResponseWriter, r *http.Request) {
 	})
 }
 
+// isAuthorized returns true if the first api token query parameter
+// matches any APIToken configuration value. Otherwise, it returns false.
+func (d *DcrtimeStore) isAuthorized(r *http.Request) bool {
+	// Extract api token from query parameters.
+	queryTokens := r.URL.Query()["apitoken"]
+
+	var apiToken string
+	if len(queryTokens) > 0 {
+		// If the request has multiple apitoken
+		// values, only consider the first one.
+		apiToken = queryTokens[0]
+	} else {
+		return false
+	}
+
+	if _, exist := d.apiTokens[apiToken]; exist {
+		return true
+	}
+
+	log.Errorf("isAuthorized %v: authentication failed", r.RemoteAddr)
+	return false
+}
+
+func (d *DcrtimeStore) walletBalance(w http.ResponseWriter, r *http.Request) {
```

Comment:
Are you leaking a body.Close here?

Reply:
Taken care of by closeBody

Reply:
Curiously, I was not able to reproduce any apparent leak by *not* closing the request body received by the server. Do you have an idea on steps to reproduce a leak here? I just want to see it break, then watch it not break w/ this latest update.

Docs suggest the server does not need to close the request body, but I'm under the assumption that I am missing something. https://github.com/go-lang/go/blob/c6f61eec333c556764ffc77d61b13ba62040f581/src/pkg/net/http/request.go#L135

w/ a dcrtimed proxy listening on 59252, this doesn't show any apparent leakiness:
$ `curl -sk --data '{"id":"dcrtime cli","digests":["b5e4f95466df4ee92a205d00d62b3704915470597e9a0f8b126374befe213bd0"]}' "https://localhost:59252/v1/verify/?[1-1000000]"`
> {"id":"dcrtime cli","digests":[{"digest":"b5e4f95466df4ee92a205d00d62b3704915470597e9a0f8b126374befe213bd0","servertimestamp":1572494400,"result":0,"chaininformation":{"chaintimestamp":0,"transaction":"0000000000000000000000000000000000000000000000000000000000000000","merkleroot":"0000000000000000000000000000000000000000000000000000000000000000","merklepath":{"NumLeaves":0,"Hashes":null,"Flags":null}}}],"timestamps":[]}

Even more fun is that not closing the proxied response body from the backend doesn't seem to leak either, so something is probably off with my testing.